topdown/trusted_identity_providers_test.go

```diff
@@ -0,0 +1,269 @@
+	OnlyGoogleIdP.Store(googleIdP, googleIdPVerifier)
+
+	var OnlyOktaIdP sync.Map
+	OnlyOktaIdP.Store(oktaIdP, oktaIdPVerifier)
+
+	var OnlyNotAnIdP sync.Map
+	OnlyNotAnIdP.Store(notAnIdP, notAnIdPVerifier)
+
+	var OnlyNotSecureIdP sync.Map
+	OnlyNotSecureIdP.Store(notSecureIdP, notSecureIdPVerifier)
+
+	var GoogleAndOktaIdP sync.Map
+	GoogleAndOktaIdP.Store(googleIdP, googleIdPVerifier)
+	GoogleAndOktaIdP.Store(oktaIdP, oktaIdPVerifier)
+
+	tests := []struct {
+		name    string
+		args    args
+		want    *TrustedIdProviderManagerImpl
+		wantErr bool
+	}{
+		{name: "Verify nil    returns empty", args: args{trustedIdentityProviders: nil},                             want: &TrustedIdProviderManagerImpl{trustedVerifiers: sync.Map{}},       wantErr: false},
+		{name: "Verify 1      returns 1",     args: args{trustedIdentityProviders: []*string{&googleIdP}},           want: &TrustedIdProviderManagerImpl{trustedVerifiers: OnlyGoogleIdP},    wantErr: false},
+		{name: "Verify new 1  returns 1",     args: args{trustedIdentityProviders: []*string{&oktaIdP}},             want: &TrustedIdProviderManagerImpl{trustedVerifiers: OnlyOktaIdP},      wantErr: false},
+		{name: "Verify 2      returns 2",     args: args{trustedIdentityProviders: []*string{&oktaIdP, &googleIdP}}, want: &TrustedIdProviderManagerImpl{trustedVerifiers: GoogleAndOktaIdP}, wantErr: false},
+		{name: "Verify 1 fake returns error", args: args{trustedIdentityProviders: []*string{&notAnIdP}},            want: &TrustedIdProviderManagerImpl{trustedVerifiers: OnlyNotAnIdP},     wantErr: true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := GetTrustedIdentityProviderManager(tt.args.trustedIdentityProviders)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("GetTrustedIdentityProviderManager() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if err != nil {
+				return
+			}
+
+			for _, idp := range allIdPs {
+				gotValue, _ := got.trustedVerifiers.Load(idp)
+				wantedValue, _ := tt.want.trustedVerifiers.Load(idp)
+				if gotValue != wantedValue {
+					t.Errorf("GetTrustedIdentityProviderManager() got = %v, but want %v", gotValue, wantedValue)
+				}
+			}
+		})
+	}
+}
+
+
+type FetchTestTokenResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType string `json:"token_type"`
+}
+func fetchLiveTestTokenFromAuth0TestAccount(t *testing.T) (*string, error){
+	url := "https://dev-wa90c9xx.auth0.com/oauth/token"
+	payload := strings.NewReader("{\"client_id\":\"0pGQ4pgCuPoH9dqFZnT7rWLL4dHU2YoS\",\"client_secret\":\"TQTFnCpbavunE5mCvivLwx0ONkphF7zh7dv5ECc_BFRyxCKhSQUIQLfFrHN_og3g\",\"audience\":\"http://opa-test-api-app/api\",\"grant_type\":\"client_credentials\"}")
+	req, err := http.NewRequest("POST", url, payload)
+	if err != nil {
+		t.Errorf("Testing account NewRequest broke, got error = %v", err)
+		return nil, err
+	}
+	req.Header.Add("content-type", "application/json")
+	res, err := http.DefaultClient.Do(req)
```

Comment:
We should try and avoid making any actual HTTP requests in the unit tests. We've set the bar that they should work in an offline environment.

Reply:
Fair point. Do we have an integration test suite somewhere?

Reply:
We don't really have one for OPA right now that would be a good spot to use external servers. What would normally happen is the tests would use the http test server https://golang.org/pkg/net/http/httptest/ and you would essentially feed mocked responses back to the test client with a handler on the test server.